step/step.go

```diff
@@ -0,0 +1,244 @@
+package step
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"slices"
+
+	"github.com/bitrise-io/go-steputils/v2/export"
+	"github.com/bitrise-io/go-steputils/v2/stepconf"
+	"github.com/bitrise-io/go-utils/retry"
+	"github.com/bitrise-io/go-utils/v2/command"
+	"github.com/bitrise-io/go-utils/v2/log"
+)
+
+const (
+	testShardsDirKey = "BITRISE_TEST_SHARDS_PATH"
+)
+
+var ErrXcodebuildFailed = errors.New("unexpected xcodebuild failure")
+
+type Input struct {
+	ProductPath      string `env:"product_path,required"`
+	ShardCount       int    `env:"shard_count,required"`
+	ShardCalculation string `env:"shard_calculation,opt[alphabetically]"`
+	Destination      string `env:"destination,required"`
+	Verbose          bool   `env:"verbose,opt[true,false]"`
+}
+
+type Config struct {
+	ProductPath      string
+	ShardCount       int
+	ShardCalculation string
+	Destination      string
+}
+
+type Result struct {
+	TestShardsDir string
+}
+
+type Step struct {
+	inputParser    stepconf.InputParser
+	commandFactory command.Factory
+	exporter       export.Exporter
+	logger         log.Logger
+}
+
+func NewStep(
+	inputParser stepconf.InputParser,
+	commandFactory command.Factory,
+	exporter export.Exporter,
+	logger log.Logger,
+) Step {
+	return Step{
+		inputParser:    inputParser,
+		commandFactory: commandFactory,
+		exporter:       exporter,
+		logger:         logger,
+	}
+}
+
+func (s *Step) ProcessConfig() (*Config, error) {
+	var input Input
+	err := s.inputParser.Parse(&input)
+	if err != nil {
+		return &Config{}, err
+	}
+
+	stepconf.Print(input)
+	s.logger.EnableDebugLog(input.Verbose)
+
+	return &Config{
+		ProductPath:      input.ProductPath,
+		ShardCount:       input.ShardCount,
+		ShardCalculation: input.ShardCalculation,
+		Destination:      input.Destination,
+	}, nil
+}
+
+func (s *Step) Run(config Config) (Result, error) {
+	s.logger.Println()
+	s.logger.Infof("Collecting tests:")
+
+	var tests []string
+
+	err := retry.Times(3).Wait(10).TryWithAbort(func(attempt uint) (error, bool) {
```

Comment:
On the newer stacks (Xcode 16.2 and edge) there might be a regression where the testmanagerd deamon crashes or hangs or something else which makes the command execution lose connection to it. This does not seem to happen on older stacks.

Based on my experiementation a simple retry could recover most of the cases but not all of them.